src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioPortableExecutableReference.FileTimeStampProvider.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+
+internal partial class VisualStudioMetadataReferenceManager
+{
+    private sealed partial class VisualStudioPortableExecutableReference
+    {
+        /// <summary>
+        /// Provides a performant mechanism to get the timestamp of a file. Caches the timestamp of requested
+        /// files and watches the directories containing those files for changes.
+        /// </summary>
+        private sealed class FileTimeStampProvider
+        {
+            private static FileTimeStampProvider? s_instance;
+
+            private readonly Dictionary<string, DateTime> _cachedTimestamps = [];
+            private readonly Dictionary<string, IFileChangeContext> _fileChangeContexts = [];
+
+            private FileTimeStampProvider(Workspace workspace)
```

Comment:
At first I was concerned that unless this was the real VisualStudioWorkspace that we might be listening to a different workspace's events. I see that's the one yuo're actually importing, but maybe it's best to keep that strongly typed here too to ensure we don't refactor it and grab something else?